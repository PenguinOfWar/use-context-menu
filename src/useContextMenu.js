import {
  useLayoutEffect,
  useRef,
  useState,
  useEffect,
  useCallback,
} from "react";
import { getMenuPosition, getRTLMenuPosition } from "./helpers";
import buildUseContextMenuTrigger from "./buildUseContextMenuTrigger";

export const keyCodes = {
  ESCAPE: 27,
  ENTER: 13,
  TAB: 9,
  UP_ARROW: 38,
  DOWN_ARROW: 40,
};
const baseStyles = {
  position: "absolute",
  opacity: 0,
  pointerEvents: "none",
  top: 0,
  left: 0,
  userSelect: "none",
  transform: "translate3d(0,0,0)",
};

const focusElement = (el) => el.focus();
const useContextMenu = ({
  rtl,
  handleElementSelect = focusElement,
  hideOnScroll = false,
} = {}) => {
  const menuRef = useRef();
  const selectables = useRef([]);
  // TODO: refactor with useReducer
  const [style, setStyles] = useState(baseStyles);
  const [selectedIndex, setSelectedIndex] = useState(-1);
  const [isVisible, setVisible] = useState(false);
  const isVisibleRef = useRef(isVisible);
  isVisibleRef.current = isVisible;
  const [coords, setCoords] = useState([0, 0]);
  const [collectedData, setCollectedData] = useState();
  const hideMenu = useCallback(() => {
    setVisible(false);
    setSelectedIndex(-1);
  }, [setVisible, setSelectedIndex]);
  const triggerVisible = useCallback(
    (coords, data) => {
      setVisible(true);
      setCoords(coords);
      setCollectedData(data);
    },
    [setVisible, setCollectedData, setCoords]
  );

  const markSelectable = (el) =>
    (selectables.current = el === null ? [] : [...selectables.current, el]);

  useEffect(() => {
    const handleOutsideClick = (e) => {
      if (
        isVisibleRef.current &&
        menuRef.current &&
        !menuRef.current.contains(e.target)
      ) {
        hideMenu();
        document.removeEventListener("mousedown", handleOutsideClick);
        document.removeEventListener("touchstart", handleOutsideClick);
      }
    };
    const handleKeyNavigation = (e) => {
      switch (e.keyCode) {
        case keyCodes.TAB:
          hideMenu();
          break;
        case keyCodes.ESCAPE:
          e.preventDefault();
          hideMenu();
          break;
        case keyCodes.UP_ARROW:
          e.preventDefault();
          if (selectedIndex > 0) {
            setSelectedIndex((s) => s - 1);
            handleElementSelect(selectables.current[selectedIndex - 1]);
          }
          break;
        case keyCodes.DOWN_ARROW:
          e.preventDefault();
          if (selectedIndex + 1 < selectables.current.length) {
            setSelectedIndex((s) => s + 1);
            handleElementSelect(selectables.current[selectedIndex + 1]);
          }
          break;
        case keyCodes.ENTER:
          if (selectedIndex !== -1) {
            selectables.current[selectedIndex].click();
          }
          hideMenu();
          break;
        default:
      }
    };
    if (isVisible) {
      document.addEventListener("mousedown", handleOutsideClick, {
        passive: true,
        capture: true,
      });
      document.addEventListener("touchstart", handleOutsideClick, {
        passive: true,
        capture: true,
      });
      hideOnScroll &&
        document.addEventListener("scroll", hideMenu, {
          once: true,
        });
      document.addEventListener("keydown", handleKeyNavigation);
    }
    return () => {
      document.removeEventListener("mousedown", handleOutsideClick);
      document.removeEventListener("touchstart", handleOutsideClick);
      document.removeEventListener("scroll", hideMenu);
      document.removeEventListener("keydown", handleKeyNavigation);
    };
  }, [
    menuRef,
    hideMenu,
    selectedIndex,
    setSelectedIndex,
    selectables,
    handleElementSelect,
    isVisible,
    hideOnScroll,
  ]);

  useLayoutEffect(() => {
    if (isVisible) {
      const rect = menuRef.current.getBoundingClientRect();
      const { top, left } = rtl
        ? getRTLMenuPosition(rect, coords)
        : getMenuPosition(rect, coords);
      setStyles((st) => ({
        ...st,
        transform: `translate3d(${left}px, ${top}px, 0)`,
        opacity: 1,
        pointerEvents: "auto",
      }));
    } else {
      setStyles(baseStyles);
    }
  }, [menuRef, isVisible, coords, rtl]);

  const bindMenuProps = {
    style,
    ref: menuRef,
    role: "menu",
    tabIndex: 0,
    onContextMenu: (e) => e.preventDefault(),
    "aria-hidden": !isVisible,
  };
  const bindMenuItemProps = {
    ref: markSelectable,
    role: "menuitem",
    tabIndex: -1,
  };
  return [
    bindMenuProps,
    bindMenuItemProps,
    buildUseContextMenuTrigger(triggerVisible),
    {
      data: collectedData,
      hideMenu,
      isVisible,
      setVisible,
      coords,
      setCoords,
    },
  ];
};

export default useContextMenu;
