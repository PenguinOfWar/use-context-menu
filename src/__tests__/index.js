import { renderHook, act } from "@testing-library/react-hooks";
import useContextMenu, { keyCodes } from "../useContextMenu";

test("useContextMenu result is correct", () => {
  const { result } = renderHook(() => useContextMenu());

  expect(result.current.length).toBe(4);

  expect(result.current).toMatchSnapshot();
});

test("useContextMenu register event listeners only when visible", () => {
  const addEventListener = jest.spyOn(document, "addEventListener");
  const { result } = renderHook(() => useContextMenu());

  result.current[0].ref.current = {
    getBoundingClientRect: jest.fn(() => ({ height: 100, width: 100 }))
  };

  expect(addEventListener).not.toHaveBeenCalled();

  act(() => {
    result.current[3].setVisible(true);
    result.current[3].setCoords([100, 100]);
  });

  expect(addEventListener).toHaveBeenCalledTimes(5);
  expect(addEventListener.mock.calls.map(call => call[0])).toEqual([
    "mousedown",
    "touchstart",
    "scroll",
    "contextmenu",
    "keydown"
  ]);
});

test("useContextMenu sets correctly the style when changing visibility", () => {
  const { result } = renderHook(() => useContextMenu());

  result.current[0].ref.current = {
    getBoundingClientRect: jest.fn(() => ({ height: 100, width: 100 }))
  };

  expect(result.current[0].style).toEqual({
    opacity: 0,
    pointerEvents: "none",
    position: "fixed"
  });

  act(() => {
    result.current[3].setVisible(true);
    result.current[3].setCoords([100, 100]);
  });

  expect(result.current[0].style).toEqual({
    left: "100px",
    opacity: 1,
    pointerEvents: "auto",
    position: "fixed",
    top: "100px"
  });

  act(() => {
    result.current[3].setVisible(false);
  });

  expect(result.current[0].style).toEqual({
    opacity: 0,
    pointerEvents: "none",
    position: "fixed"
  });
});

test("useContextMenu keeps the menu inside the viewport", () => {
  const { result } = renderHook(() => useContextMenu());

  result.current[0].ref.current = {
    getBoundingClientRect: jest.fn(() => ({ height: 100, width: 100 }))
  };

  expect(result.current[0].style).toEqual({
    opacity: 0,
    pointerEvents: "none",
    position: "fixed"
  });

  act(() => {
    result.current[3].setVisible(true);
    result.current[3].setCoords([1000, 700]);
  });

  expect(result.current[0].style).toEqual({
    left: "900px",
    opacity: 1,
    pointerEvents: "auto",
    position: "fixed",
    top: "600px"
  });
});

test("useContextMenu hides the menu when a new contextmenu event is triggered", () => {
  const { result } = renderHook(() => useContextMenu());

  result.current[0].ref.current = {
    getBoundingClientRect: jest.fn(() => ({ height: 100, width: 100 }))
  };

  act(() => {
    result.current[3].setVisible(true);
  });

  act(() => {
    const event = new MouseEvent("contextmenu");

    document.dispatchEvent(event);
  });

  expect(result.current[3].isVisible).toEqual(false);
});

test("useContextMenu hides the menu when clicking outside", () => {
  const { result } = renderHook(() => useContextMenu());

  result.current[0].ref.current = {
    getBoundingClientRect: jest.fn(() => ({ height: 100, width: 100 })),
    contains: () => false
  };

  act(() => {
    result.current[3].setVisible(true);
  });

  act(() => {
    const event = new MouseEvent("mousedown");

    document.dispatchEvent(event);
  });

  expect(result.current[3].isVisible).toEqual(false);
});

test("useContextMenu hides the menu when clicking outside", () => {
  const { result } = renderHook(() => useContextMenu());

  result.current[0].ref.current = {
    getBoundingClientRect: jest.fn(() => ({ height: 100, width: 100 })),
    contains: () => false
  };

  act(() => {
    result.current[3].setVisible(true);
  });

  act(() => {
    const event = new MouseEvent("touchstart");

    document.dispatchEvent(event);
  });

  expect(result.current[3].isVisible).toEqual(false);
});

test("useContextMenu hides the menu on scroll", () => {
  const { result } = renderHook(() => useContextMenu());

  result.current[0].ref.current = {
    getBoundingClientRect: jest.fn(() => ({ height: 100, width: 100 })),
    contains: () => false
  };

  act(() => {
    result.current[3].setVisible(true);
  });

  act(() => {
    document.dispatchEvent(new Event("scroll"));
  });

  expect(result.current[3].isVisible).toEqual(false);
});

test("useContextMenu hides the menu on ESCAPE key press", () => {
  const { result } = renderHook(() => useContextMenu());

  result.current[0].ref.current = {
    getBoundingClientRect: jest.fn(() => ({ height: 100, width: 100 })),
    contains: () => false
  };

  act(() => {
    result.current[3].setVisible(true);
  });

  act(() => {
    const event = new KeyboardEvent("keydown", { keyCode: keyCodes.ESCAPE });
    document.dispatchEvent(event);
  });

  expect(result.current[3].isVisible).toEqual(false);
});

test("useContextMenu hides the menu on ENTER key press", () => {
  const { result } = renderHook(() => useContextMenu());

  result.current[0].ref.current = {
    getBoundingClientRect: jest.fn(() => ({ height: 100, width: 100 })),
    contains: () => false
  };

  act(() => {
    result.current[3].setVisible(true);
  });

  act(() => {
    const event = new KeyboardEvent("keydown", { keyCode: keyCodes.ENTER });
    document.dispatchEvent(event);
  });

  expect(result.current[3].isVisible).toEqual(false);
});

test("useContextMenu index of item selected by keyboard gets reset when closing the menu", () => {
  let selectedEl = 0;
  const { result } = renderHook(() =>
    useContextMenu({
      handleElementSelect: _selectedEl => (selectedEl = _selectedEl)
    })
  );

  result.current[0].ref.current = {
    getBoundingClientRect: jest.fn(() => ({ height: 100, width: 100 })),
    contains: () => false
  };

  result.current[1].ref(1);
  result.current[1].ref(2);
  result.current[1].ref(3);

  act(() => {
    result.current[3].setVisible(true);
  });

  act(() => {
    const event = new KeyboardEvent("keydown", { keyCode: 40 });

    document.dispatchEvent(event);
  });

  act(() => {
    const event = new MouseEvent("touchstart");

    document.dispatchEvent(event);
  });

  act(() => {
    result.current[3].setVisible(true);
  });

  act(() => {
    const event = new KeyboardEvent("keydown", { keyCode: 40 });

    document.dispatchEvent(event);
  });

  expect(selectedEl).toEqual(1);
});
