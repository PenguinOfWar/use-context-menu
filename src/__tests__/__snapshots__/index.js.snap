// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useContextMenu result is correct 1`] = `
Array [
  Object {
    "ref": Object {
      "current": undefined,
    },
    "role": "menu",
    "style": Object {
      "opacity": 0,
      "pointerEvents": "none",
      "position": "fixed",
    },
    "tabIndex": -1,
  },
  Object {
    "ref": [Function],
    "role": "menuitem",
    "tabIndex": -1,
  },
  [Function],
  Object {
    "coords": Array [
      0,
      0,
    ],
    "data": undefined,
    "isVisible": false,
    "setCoords": [Function],
    "setVisible": [Function],
  },
]
`;
